lib/srv/alpnproxy/aws_local_proxy.go

```diff
@@ -47,8 +47,14 @@ func (m *AWSAccessMiddleware) CheckAndSetDefaults() error {
 }
 
 func (m *AWSAccessMiddleware) HandleRequest(rw http.ResponseWriter, req *http.Request) bool {
-	if err := aws.VerifyAWSSignature(req, m.AWSCredentials); err != nil {
-		m.Log.WithError(err).Error("AWS signature verification failed.")
+	// It's found that some AWS clients may have slightly different signing
+	// logic than the Go SDK. For example, the Java SDK v1 includes the
+	// "Use-Agent" header when signing but the Go SDK explicitly excludes this
+	// header. To workaround, only AccessKeyID from the signature is verified
```

Comment:
Could we try to verify both with and without the User-Agent? Would it provide any hints to whether we are seeing a request from Go or Java?

Reply:
> Could we try to verify both with and without the User-Agent? 
Some of the signature logic including [`ignoredHeaders` ](https://github.com/aws/aws-sdk-go/issues/4662) is private and hard-coded in the Go SDK. So we cannot try to sign "with" it.

> Would it provide any hints to whether we are seeing a request from Go or Java?
It's kinda possible. We know what headers are included in the signature. So technically we can verify full when 'User-Agent' is not present there, and only do the AccessKeyID way only when 'User-Agent' is present. Do you think that would be better? My thought process was if we don't verify the whole thing for some cases, then why bother doing it for others. But I am open to preserving the full check if we think it's better for security.

This also ties to your question about dropping the signature validation. As mentioned in the description, the local validation is optional at best. We don't do validations for other local proxies (e.g `tsh proxy db --tunnel`).

Reply:
> Some of the signature logic including https://github.com/aws/aws-sdk-go/issues/4662is private and hard-coded in the Go SDK. So we cannot try to sign "with" it.

I thought you might say that.

> So technically we can verify full when 'User-Agent' is not present there, and only do the AccessKeyID way only when 'User-Agent' is present

If signature validation serves a purpose, I think I'd rather do it only for Go than not at all. If it doesn't serve a purpose, then I'd be OK cutting it out. I don't understand the overall picture to make a call, though.

Reply:
Ok. Then let me update this part to only do AccessKeyID validation when 'User-Agent' is getting signed. So everything else is same as before